### ISSUE TITLE
[Design Doc] for /notify Command

### ISSUE THREAD

**Issue by sahsisunny** [Real-Dev-Squad/discord-slash-commands#125]
# Design Doc for `/notify` Command

## Overview

The `/notify` command is designed to provide task deadline notifications and onboarding status monitoring for our Discord bot. This document outlines the technical details and implementation guidelines for this command.

## User Stories

- As an ADMIN/SUPER USER, I want to be able to use the `/notify` command to quickly identify users with overdue tasks.
- As an ADMIN/SUPER USER, I want to monitor the onboarding progress of new members based on predefined criteria.

## Command Structure

The `/notify` command will accept two primary options: "OVERDUE" and "ONBOARDING," each with their respective sub-options.

### Option 1: OVERDUE

/notify OVERDUE: In 1 Day
/notify OVERDUE: In 3 Days

### Option 2: ONBOARDING

/notify ONBOARDING: > 31 Days
/notify ONBOARDING: > 7D & < 31D

### Command Parsing

- The bot will listen for messages that start with `/notify`.
- It will parse the command to determine the selected option (OVERDUE or ONBOARDING) and any sub-options.

#### OVERDUE

- When the user selects the OVERDUE option, the bot will provide the sub options if the user specifies a time frame (e.g., "In 1 Day"), the bot will calculate the deadline date and filter tasks accordingly.

#### ONBOARDING

- When the user selects the ONBOARDING option, the bot will provide the sub options if the user It will compare the onboarding date with the specified criteria to identify users falling into the selected category.

### Error Handling
- There is no need to any custom input query so there is no chance to handle any errors.

### Response

- The bot's response will include a list of users or tasks that match the selected criteria.
- The response will be formatted in a clear and user-friendly manner.

Ex:
```
    **Message**

    @user1 @user2 @user3 @user4
    @user5 @user6 @user7 @user8
```
## Testing

- Comprehensive testing will be performed to ensure the command's correctness and reliability.
- Test cases will cover various scenarios.

## Documentation

- Clear documentation will be provided to users on how to use the `/notify` command and its available options.
- Documentation will include examples of valid command usage.

## Timeline

This design will be implemented as a part of the bot's functionality. The timeline for implementation will be determined based on project priorities.

**Comment by j24m** [comment]
LGTM

**Comment by shreya-mishra** [comment]
LGTM

**Comment by prakashchoudhary07** [comment]
In the sequence diagram, for overdue you have mentioned that the discord bot will get task Data, calculate the deadline, and filter overdue tasks Similarly for onboarding also discord bot will receive user Data, and compare it with criteria and other calculations, How will it do it exactly?
Wanted to ask how will discord bot get those Will it connect to the backend or Clouflare workers or Firestore?

**Comment by prakashchoudhary07** [comment]
```
Error Handling

    There is no need to any custom input query so there is no chance to handle any errors.

```

I didn't understand what this exactly meant. Why is there no chance to handle errors?

**Comment by prakashchoudhary07** [comment]
```
User Stories

    As an ADMIN/SUPER USER, I want to be able to use the /notify command to quickly identify users with overdue tasks.
    As an ADMIN/SUPER USER, I want to monitor the onboarding progress of new members based on predefined criteria.
```

Just clearing things up, the notify command will be used to notify users right? 
Or identify users, if yes then how?

Can you please elaborate second point please

**Comment by sahsisunny** [comment]
> ```
> Error Handling
> 
>     There is no need to any custom input query so there is no chance to handle any errors.
> ```
> 
> I didn't understand what this exactly meant. Why is there no chance to handle errors?

We are not taking any input from the user.

**Comment by sahsisunny** [comment]
> ```
> User Stories
> 
>     As an ADMIN/SUPER USER, I want to be able to use the /notify command to quickly identify users with overdue tasks.
>     As an ADMIN/SUPER USER, I want to monitor the onboarding progress of new members based on predefined criteria.
> ```
> 
> Just clearing things up, the notify command will be used to notify users right? Or identify users, if yes then how?
> 
> Can you please elaborate second point please



> ```
> User Stories
> 
>     As an ADMIN/SUPER USER, I want to be able to use the /notify command to quickly identify users with overdue tasks.
>     As an ADMIN/SUPER USER, I want to monitor the onboarding progress of new members based on predefined criteria.
> ```
> 
> Just clearing things up, the notify command will be used to notify users right? Or identify users, if yes then how?
> 
> Can you please elaborate second point please

It gives the list/tag in a group(like mention each commend) of users based on conditions.

**Comment by sahsisunny** [comment]
> This command will notify a group of users at once right? Wouldn't it be better to notify them individually?

Yes, you're right. This command is designed to notify a group of users all at once. However, I'd like to understand more about the scenario where individual notifications would be beneficial. Could you please provide more details on the use case for notifying users individually? This would help me better understand your requirements and explore any potential improvements to the notification process.

**Comment by prakashchoudhary07** [comment]
I just wanted to ask if we will get all tasks and onboarding data, and filter at the backend. or the DB will give the filtered results

**Comment by prakashchoudhary07** [comment]
Nice, LGTM
